airbyte-integrations/connectors/destination-oracle/src/main/java/io/airbyte/integrations/destination/oracle/OracleOperations.java

```diff
@@ -0,0 +1,205 @@
+   */
+
+  private boolean tableExists(JdbcDatabase database, String schemaName, String tableName) throws Exception {
+    Integer count = database.queryInt("select count(*) \n from all_tables\n where upper(owner) = upper(?) and upper(table_name) = upper(?)",
+        schemaName, tableName);
+    return count == 1;
+  }
+
+  @Override
+  public void dropTableIfExists(JdbcDatabase database, String schemaName, String tableName) throws Exception {
+    if (tableExists(database, schemaName, tableName)) {
+      try {
+        final String query = String.format("DROP TABLE %s.%s", schemaName, tableName);
+        database.execute(query);
+      } catch (Exception e) {
+        LOGGER.error(String.format("Error dropping table %s.%s", schemaName, tableName), e);
+        throw e;
+      }
+    }
+  }
+
+  @Override
+  public String truncateTableQuery(String schemaName, String tableName) {
+    return String.format("DELETE FROM %s.%s\n", schemaName, tableName);
+  }
+
+  @Override
+  public void insertRecords(JdbcDatabase database, List<AirbyteRecordMessage> records, String schemaName, String tempTableName)
+      throws Exception {
+    final String tableName = String.format("%s.%s", schemaName, tempTableName);
+    final String columns = String.format("(%s, %s, %s)",
+        JavaBaseConstants.COLUMN_NAME_AB_ID.substring(1),
+        JavaBaseConstants.COLUMN_NAME_DATA.substring(1),
+        JavaBaseConstants.COLUMN_NAME_EMITTED_AT.substring(1));
+    final String recordQueryComponent = "(?, ?, ?)\n";
+    insertRawRecordsInSingleQuery(tableName, columns, recordQueryComponent, database, records, UUID::randomUUID);
+  }
+
+  static void insertRawRecordsInSingleQuery(String tableName,
+                                            String columns,
+                                            String recordQueryComponent,
+                                            JdbcDatabase jdbcDatabase,
+                                            List<AirbyteRecordMessage> records,
+                                            Supplier<UUID> uuidSupplier)
+      throws SQLException {
+    if (records.isEmpty()) {
+      return;
+    }
+
+    jdbcDatabase.execute(connection -> {
+
+      // Strategy: We want to use PreparedStatement because it handles binding values to the SQL query
+      // (e.g. handling formatting timestamps). A PreparedStatement statement is created by supplying the
+      // full SQL string at creation time. Then subsequently specifying which values are bound to the
+      // string. Thus there will be two loops below.
+      // 1) Loop over records to build the full string.
+      // 2) Loop over the records and bind the appropriate values to the string.
+      final StringBuilder sql = new StringBuilder("INSERT ALL ");
+      records.forEach(r -> sql.append(String.format("INTO %s %s VALUES %s", tableName, columns, recordQueryComponent)));
+      sql.append(" SELECT 1 FROM DUAL");
```

Comment:
This is how you do a multi-row insert in Oracle.  It's a really ugly hack: It's "supposed to" look like `insert all into table_a select * from table_b`, but you're allowed to use syntax similar to the standard `INSERT` within there.  But there is a formal requirement that it must end with a SELECT statement; if I don't put line 165 in there, it will error out.  So I put in essentially a dummy select.  (In Oracle, it's another formal requirement that every SELECT must come FROM a table/view/whatever.  `DUAL` is a dummy table that exists entirely to satisfy this requirement for queries that don't really need it.  In any other database that would just be `SELECT 1`.)

Reply:
can you add this explanation as a comment please? at least for me if i ever have to debug this it'll be helpful to know!